### PR TITLE
Inferable quantizers

### PR DESCRIPTION
**Pull request by reuvenperetz**
Added inferable quantizers to the quantization infrastructure:
BaseInferableQuantizer - class for a basic quantizer for inference only.

BaseTrainableQuantizer(BaseInferableQuantizer) - class for a basic quantizers for training.
BaseKerasTrainableQuantizer(BaseTrainableQuantizer) - class for a basic quantizers for training in Keras.
BasePytorchTrainableQuantizer(BaseTrainableQuantizer) - basic quantizer training class in Pytorch.

From BaseInferableQuantizer inherits now two new classes: BasePytorchInferableQuantizer and BaseKerasInferableQuantizer.
Each FW quantizer in the exportable model (aka fully quantized model) should implement one of these inferable quantizers.

From BaseKerasInferableQuantizer inherits:
BasePOTInferableQuantizer
BaseSymmetricInferableQuantizer
BaseUniformInferableQuantizer

and a weights quantizer and an activation quantizer inherit them (two for each method).

### CODE REVIEW

File: model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/inferable_quantizer_factory.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+from model_compression_toolkit.qunatizers_infrastructure.keras.inferable_quantizers.pot_inferable_quantizer import \
+    POTInferableQuantizer
+from model_compression_toolkit.qunatizers_infrastructure.keras.inferable_quantizers.symmetric_inferable_quantizer \
+    import \
+    SymmetricInferableQuantizer
+from model_compression_toolkit.qunatizers_infrastructure.keras.inferable_quantizers.uniform_inferable_quantizer import \
+    UniformInferableQuantizer
+
+
+QUANTIZATION_METHOD_2_QUANTIZER = {QuantizationMethod.POWER_OF_TWO: POTInferableQuantizer,
```
Comment (haihabi) [inline model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/inferable_quantizer_factory.py:26]: Why do we need a factory?

File: model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/pot_inferable_quantizer.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+import numpy as np
+
+from model_compression_toolkit.qunatizers_infrastructure.common.base_inferable_quantizer import QuantizationTarget
+from model_compression_toolkit.qunatizers_infrastructure.keras.inferable_quantizers.symmetric_inferable_quantizer \
+    import \
+    SymmetricInferableQuantizer
+
+
+class POTInferableQuantizer(SymmetricInferableQuantizer):
```
Comment (haihabi) [inline model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/pot_inferable_quantizer.py:24]: We not have a sperate quantizer for activation and weights ?

File: model_compression_toolkit/exporter/model_wrapper/keras/builder/fully_quantized_model_builder.py [hunk 3]
```diff
@@ -14,36 +14,29 @@
...
-    is_layer_fake_quant, get_node_name_from_layer
-from model_compression_toolkit.core.keras.quantizer.input_layer_quantize_transform import InputLayerWrapperTransform
+from model_compression_toolkit.core.common import Graph
+from model_compression_toolkit.core.keras.back2framework.keras_model_builder import KerasModelBuilder
+from model_compression_toolkit.exporter.model_wrapper.keras.builder.node_to_dispatcher import \
+    get_quantization_dispatcher
+
+
+def _get_wrapper(node: common.BaseNode,
+                 layer: Layer):
```
Comment (eladc-git) [inline model_compression_toolkit/exporter/model_wrapper/keras/builder/fully_quantized_model_builder.py:28]: output type

File: model_compression_toolkit/exporter/model_wrapper/keras/builder/node_to_quantizer.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        quantization_method = node_qc.activation_quantization_method
+        assert quantization_method in QUANTIZATION_METHOD_2_ACTIVATION_QUANTIZER, f'{quantization_method} for weights ' \
+                                                                                  f'not in ' \
+                                                                                  f'supported quantization methods ' \
+                                                                                  f'for inferable' \
+                                                                                  f' quantizers'
+        if quantization_method in [QuantizationMethod.POWER_OF_TWO,
+                                   QuantizationMethod.SYMMETRIC]:
+            return {'num_bits': node_qc.activation_n_bits,
+                    'threshold': node_qc.activation_quantization_params.get(THRESHOLD),
```
Comment (eladc-git) [inline model_compression_toolkit/exporter/model_wrapper/keras/builder/node_to_quantizer.py:70]: What about PO2 bool flag? We don't want the convertor to do some calculations on threshold to check it? same for weights quantizers.

File: model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/activation_inferable_quantizers/activation_pot_inferable_quantizer.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+from model_compression_toolkit.qunatizers_infrastructure.common.base_inferable_quantizer import QuantizationTarget
+
+
+
+if FOUND_TF:
+    import tensorflow as tf
+    from model_compression_toolkit.qunatizers_infrastructure.keras.inferable_quantizers.base_pot_inferable_quantizer \
+        import \
+        BasePOTInferableQuantizer
+    class ActivationPOTInferableQuantizer(BasePOTInferableQuantizer):
```
Comment (eladc-git) [inline model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/activation_inferable_quantizers/activation_pot_inferable_quantizer.py:28]: Why unique quantizer for PO2? We implemented it as Symmetric with flag. You can see it look pretty much like symmetric_inferable_quantizer

File: model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/base_pot_inferable_quantizer.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+import numpy as np
+
+from model_compression_toolkit.qunatizers_infrastructure.common.base_inferable_quantizer import QuantizationTarget
+from model_compression_toolkit.qunatizers_infrastructure.keras.inferable_quantizers.base_keras_inferable_quantizer \
+    import \
+    BaseKerasInferableQuantizer
+
+
+class BasePOTInferableQuantizer(BaseKerasInferableQuantizer):
```
Comment (eladc-git) [inline model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/base_pot_inferable_quantizer.py:24]: same here. lack of reuse code of symmetric.

File: tests/keras_tests/feature_networks_tests/feature_networks/scale_equalization_test.py [hunk 7]
```diff
@@ -117,6 +117,15 @@ def compare(self, quantized_model, float_model, input_x=None, quantization_info=
                             axis=first_layer_axes)
             beta = np.mean(float_model_layer2_weight / quantized_model_layer2_weight, axis=second_layer_axes)
 
+        print('*****************')
+        print(f'alpha: {alpha}')
+        print('*****************')
+        print(f'beta: {beta}')
+        print('*****************')
```
Comment (eladc-git) [inline tests/keras_tests/feature_networks_tests/feature_networks/scale_equalization_test.py:124]: How this related to this PR?

File: model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/base_pot_inferable_quantizer.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        self.num_bits = num_bits
+        self.threshold = threshold
+        self.signed = signed
+
+        is_threshold_pot = np.all([int(np.log2(x)) == np.log2(x) for x in threshold.flatten()])
+        assert is_threshold_pot, f'Expected threshold to be power of 2 but is {threshold}'
+
+        delta = threshold / (2 ** (num_bits - int(signed)))
+        self.min_range = -threshold if signed else 0
+        self.max_range = threshold - delta
```
Comment (eladc-git) [inline model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/base_pot_inferable_quantizer.py:44]: Where is get config function?

File: model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/base_keras_inferable_quantizer.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+class BaseKerasInferableQuantizer(BaseInferableQuantizer):
+    def __init__(self,
+                 quantization_target: QuantizationTarget):
+        """
+        This class is a base quantizer which validates the provided quantization config and defines an abstract
+        function which any quantizer needs to implement.
+
+        Args:
+            quantization_target: A enum which selects the quantizer tensor type: activation or weights.
+        """
```
Comment (eladc-git) [inline model_compression_toolkit/qunatizers_infrastructure/keras/inferable_quantizers/base_keras_inferable_quantizer.py:29]: Where is get config function as abstract?

File: model_compression_toolkit/qunatizers_infrastructure/common/base_inferable_quantizer.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+# ==============================================================================
+from enum import Enum
+
+
+class QuantizationTarget(Enum):
+    Activation = 0
+    Weights = 1
+
+
+class BaseInferableQuantizer:
```
Comment (eladc-git) [inline model_compression_toolkit/qunatizers_infrastructure/common/base_inferable_quantizer.py:23]: I think this name should be BaseQuantizer without the infereable, and BaseTrainableQuantizer and BaseInferableQuantizer should inherit from it.

**Review by eladc-git — CHANGES_REQUESTED** [review]

**Comment by reuvenperetz** [comment]
Two things:
1. There is a fix for a non-related issue (in scale equalization). It will be merged in a different PR (this is just for the tests to pass).
2. The design hierarchies are supposed to keep it simple (this is why for example, I choose not to inherit quantizers from different quantization methods one from another (pot -> symmetric -> uniform). If you find it problematic, we can talk about it.

**Review by eladc-git — APPROVED** [review]

**Pull request merged**